### ISSUE TITLE
chrome 106 Ja3 fails with error: remote error: tls: unexpected message map[]

### ISSUE THREAD

**Issue by sla-te** [Danny-Dasilva/CycleTLS#169]
go 1.19

### Description

```
func main() {

	client := cycletls.Init()
	response, err := client.Do("https://tools.scrapfly.io/api/fp/ja3?extended=1", cycletls.Options{
		Body:      "",
		Ja3:       "772,4865-4866-4867-49195-49199-49196-49200-52393-52392-49171-49172-156-157-47-53,0-23-65281-10-11-35-16-5-13-18-51-45-43-27-17513-41,29-23-24,0",
		UserAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36",
	}, "GET")
	if err != nil {
		log.Print("Request Failed: " + err.Error())
	}
	log.Println(response)
}
```

### Issue Type

Bug

### Operating System

Windows 10

### Golang Version

Other

### Relevant Log Output

```shell
Get "https://tools.scrapfly.io/api/fp/ja3?extended=1": uTlsConn.Handshake() error: remote error: tls: unexpected message map[]}
```

**Comment by 89z** [comment]
772 is never a valid number for TLS

**Comment by kasiusm** [comment]
> 772 is never a valid number for TLS
???

`772` is `0x0304` in hex, https://tlsfingerprint.io/top/versions it's tls 1.3 https://datatracker.ietf.org/doc/html/draft-ietf-tls-tls13-22

**Comment by 89z** [comment]
> `772` is `0x0304` in hex, https://tlsfingerprint.io/top/versions it's tls 1.3 https://datatracker.ietf.org/doc/html/draft-ietf-tls-tls13-22

Yes thats all true, but 772 is never a valid value for the `SSLVersion` argument of a JA3. Over the wire, the TLS version is actually set in three places:

1. record version
2. handshake version
3. supported versions

with regards to JA3, "handshake version" should always be used. "supported versions" cannot be used, as its an array not a single value. further, "record version" cannot be used, as its deprecated since TLS 1.3. More info:

https://datatracker.ietf.org/doc/html/rfc8446#appendix-D

In summary, the TLSVersion should never be higher than `0x0303` (771), as its locked at that value since TLS 1.3. That means that anyone reporting 772 or higher is wrong.

**Comment by kasiusm** [comment]
for database having reference based on this, yes it's a good thing to keep it consistent, but newer tools should adapt,
 this field is irrelevant for fingerprint if it's frozen. Specially if the usage is to bring back settings like this here, this lib take a ja3 as configuration, so tls version is restored from it, freeze to 771 produce wrong settings. 
 
>  the TLS version should never be higher than 0x0303 (771)

This is an implementation detail of TLS protocol, for BC reason it's been decided to freeze it - the initial goal of ja3 was to fingerprint the negociated version - ja3 spec are old and no new version has been drafted and IMO there is no reason for us to follow TLS spec that state "TLS version should never be higher than 0x0303 (771)", the goal is to produce coherent/immutable fingerprint that reflect the TLS settings used
 
 > That means anyone reporting 772 or higher is wrong.
 
 Actually, reporting 771 for tls1.3 is also wrong, whatever the side you are, as fingerprinter you report the wrong version and you miss the data point of the negociated version (since it's freeze) and as WAF bypass, setup TLS from ja3 is also will also produce wrong settings. So yes there is a free interpretation of SSLVersion

**Comment by 89z** [comment]
> newer tools should adapt

No, they should not. JA3 is already specified [1], as using a single value, the handshake version as previously mentioned. If anything 772 or higher is used, *then its no longer a JA3, its something else*. If you want to start a spec called KASIUSM fingerprint, I would support that. But it wont be JA3.

> this field is irrelevant for fingerprint if it's frozen.

Yes I agree. Ideally the `SSLVersion` JA3 field should instead report TLS `supported_versions` [2] commonly found under extension 43, so maybe something like this:

~~~
771-772,other-stuff
~~~

> the initial goal of ja3 was to fingerprint the negociated version

No, this has never been the case. Even from the first version of the spec [3], only the single value of handshake version was used.

> ja3 spec are old and no new version has been drafted and IMO there is no reason for us to follow TLS spec that state "TLS version should never be higher than 0x0303 (771)", the goal is to produce coherent/immutable fingerprint that reflect the TLS settings used

I think thats a good goal. However you cannot use JA3 for that goal, as JA3 `SSLVersion` value is locked to something other than what you want it to be. If you wish to start some new fingerprint for TLS, then you are welcome to do so.

> Actually, reporting 771 for tls1.3 is also wrong, whatever the side you are, as fingerprinter you report the wrong version and you miss the data point of the negociated version (since it's freeze) and as WAF bypass, setup TLS from ja3 will also produce wrong settings. So yes there is a free interpretation of SSLVersion

The spec might be ambiguous, but the implementation is not. It is and always been be a single value, the handshake version. Ideally the value for TLS `supported_versions` extension would be used instead, but thats not the decision that was made, and you are not in a position to change it, however distasteful you might find it to be.

1. https://github.com/salesforce/ja3/blob/382cd37e/python/ja3.py#L214
2. https://iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml
3. https://github.com/salesforce/ja3/blob/d61c40f7/python/ja3.py#L133

**Comment by 89z** [comment]
https://github.com/Danny-Dasilva/CycleTLS/issues/160

**Comment by Danny-Dasilva** [comment]
So maybe there's some confusion, the goal of this project is not to create a string representing tls sessions, but to spoof existing ja3 tokens. If you need to specify tls 1.3 only (which I don't think a lot of servers accept because you need to renegotiate after the initial request to get tls 1.3) then I can add a param for that. This package did not define the ja3 spec, it is merely a tool for spoofing the associated client request.
